File: kotlin/exercises/src/main/java/io/github/ajoz/workshop/fp/java/part_2/solutions/exercise_6/Exercise6.java
package io.github.ajoz.workshop.fp.java.part_2.solutions.exercise_6;

import io.github.ajoz.workshop.fp.java.tools.Function1;

import java.util.ArrayList;
import java.util.List;

class Exercise6 {
    // Part 1:
    static <A, B> Function1<B, Function1<List<A>, B>> foldLeftCurried(final Function1<B, Function1<A, B>> operator) {
        return (B initial) -> (List<A> list) -> {
            B accumulator = initial;
            for (final A element : list) {
                accumulator = operator.apply(accumulator).apply(element);
            }

            return accumulator;
        };
    }

    // Part 2:
    static Function1<List<Integer>, Integer> sum =
            foldLeftCurried((Integer a) -> (Integer b) -> a + b).apply(0);

    static Function1<List<Integer>, Integer> product =
            foldLeftCurried((Integer a) -> (Integer b) -> a + b).apply(1);

    // Part 3:
    static <A, B> Function1<List<A>, List<B>> mapCurried(final Function1<A, B> mapper) {
        return (List<A> list) -> {
            final List<B> mapped = new ArrayList<>(list.size());
            for (final A a : list) {
                mapped.add(mapper.apply(a));
            }
            return mapped;
        };
    }

    // Part 4:
    static Function1<List<String>, List<Integer>> lengths =
            mapCurried(String::length);
}

File: kotlin/tools/src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/part_2/solutions/exercise_6/Exercise6.kt
@file:Suppress("PackageName")

package io.github.ajoz.workshop.fp.kotlin.part_2.solutions.exercise_6

import java.util.*

object Exercise6 {

    // Part 1:
    fun <A, B> foldLeftCurried(operator: (B) -> (A) -> B): (B) -> (List<A>) -> B {
        return { initial: B ->
            { list: List<A> ->
                var accumulator = initial
                for (element in list) {
                    accumulator = operator(accumulator)(element)
                }

                accumulator
            }
        }
    }

    // Part 2:
    var sum = foldLeftCurried { a: Int -> { b: Int -> a + b } }(0)

    var product = foldLeftCurried { a: Int -> { b: Int -> a + b } }(1)

    // Part 3:
    fun <A, B> mapCurried(mapper: (A) -> B): (List<A>) -> List<B> {
        return { list: List<A> ->
            val mapped = ArrayList<B>(list.size)
            for (a in list) {
                mapped.add(mapper(a))
            }
            mapped
        }
    }

    // Part 4:
    var lengths = mapCurried { str: String -> str.length }
}